modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/EventSubscriber/WebtoolsAnalyticsRulesEventSubscriber.php

```diff
@@ -0,0 +1,110 @@
+use Drupal\Core\Entity\EntityTypeInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\oe_webtools_analytics\Event\AnalyticsEvent;
+use Drupal\oe_webtools_analytics\AnalyticsEventInterface;
+use Drupal\oe_webtools_analytics_rules\RuleMatcherInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Event subscriber for the Webtools Analytics event.
+ */
+class WebtoolsAnalyticsRulesEventSubscriber implements EventSubscriberInterface {
+
+  /**
+   * The rule matcher.
+   *
+   * @var \Drupal\oe_webtools_analytics_rules\RuleMatcherInterface
+   */
+  protected $ruleMatcher;
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * WebtoolsAnalyticsEventSubscriber constructor.
+   *
+   * @param \Drupal\oe_webtools_analytics_rules\RuleMatcherInterface $ruleMatcher
+   *   The rule matcher.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity type manager.
+   */
+  public function __construct(RuleMatcherInterface $ruleMatcher, EntityTypeManagerInterface $entityTypeManager) {
+    $this->ruleMatcher = $ruleMatcher;
+    $this->entityTypeManager = $entityTypeManager;
+  }
+
+  /**
+   * Event listener that returns the rule that matches the current path.
+   *
+   * When an AnalyticsEvent is fired this will match the current path against
+   * all Webtools Analytics rules that are defined and set the resulting site
+   * section on the event if a match is found.
+   *
+   * @param \Drupal\oe_webtools_analytics\AnalyticsEventInterface $event
+   *   The analytics event.
+   */
+  public function onAnalyticsEvent(AnalyticsEventInterface $event): void {
+    // We return results based on rule entities. This means that if a rule is
+    // added or deleted, or if any of the existing rules change, the cached
+    // results should be invalidated.
+    $webtools_analytics_rule_definition = $this->getWebtoolsAnalyticsRuleDefinition();
+    $webtools_rules_cache_tags = $webtools_analytics_rule_definition->getListCacheTags();
+    $event->addCacheTags($webtools_rules_cache_tags);
+    $event->addCacheContexts($webtools_analytics_rule_definition->getListCacheContexts());
+
+    // Since the rules that are used to discover the site sections are URI based
+    // the result cache should vary based on the path.
+    $event->addCacheContexts(['url.path']);
+
+    // Store the section that matches the current path on the event if it is'
+    // found.
+    $section = $this->ruleMatcher->getMatchingSection();
+    if ($section) {
+      $event->setSiteSection($section);
+    }
+  }
+
+  /**
+   * Returns the entity type definition for the Webtools Analytics Rule entity.
+   *
+   * @return \Drupal\Core\Entity\EntityTypeInterface
+   *   The entity type definition.
+   */
+  protected function getWebtoolsAnalyticsRuleDefinition(): EntityTypeInterface {
+    try {
```

Comment:
Welllll.... I think this is not a questionable practice at all but a very useful feature that trickled down from Java to PHP and is leveraged effectively by the exception classes in Drupal core.I think this feature is sadly overused and it's very positive that this is being pointed out by PHPStorm. If this would not be useful then core could just throw unchecked `RuntimeExceptions` for everything. But they are not, except when the error is a runtime condition which can not be dealt with higher up the stack.

The whole idea behind checked exceptions is that they _have to be dealt with_. Stricter languages like Java won't even compile if there are any checked exceptions allowed to bubble up to the top of the call stack. 

For this particular piece of code this is the right spot for handling these particular exceptions. We are calling into a dynamic system that will throw checked exceptions in case invalid input is given. But we know for a fact that we are passing valid data since we define the entity type ourselves. We can safely convert the exception so the calling code does not need to be concerned about handling these exceptions. After all why would an event dispatcher even be concerned with catching checked exceptions for the entity type manager?

Reply:
The value is that we are responsibly handling checked exceptions. The `catch (PluginNotFoundException $e)` is the whole point of this code. The conversion into a `\RuntimeException` is just a small bonus. But since we are calling into code that throws checked exceptions it is our responsibility to catch all of them.

I tried to explain this in my previous comment but it doesn't seem to be clear. I'll elaborate a bit.

**Checked vs unchecked**

Exceptions are thrown when an error occurs in the code which is outside of the scope of what that particular method is designed to deal with. There are two types of problems for which exceptions are thrown: either the problem is something that the calling code can deal with (e.g. trying to load a user with a name that doesn't exist), or it is something completely out of control of the codebase (e.g. the disk is full, or the database server is down). The first type is called a "checked exception" (because it should be "checked" (caught) by the calling code) and the second is an "unchecked exception" since it cannot be handled and should bubble up to the root of the call stack or the generic error handler.

PHP doesn't have the concept of checked exceptions built in to the language, but by convention all exceptions except `\RuntimeException` are considered to be checked exceptions. This is in fact identical to how it is handled in Java, except that PHP will not throw errors at compile time. Luckily we can detect them with static analysis. PHPStorm (being based on IntelliJ) has very good support for it.

Drupal core is (for the most part) correct in their use of emitting the right exception type. For example the entity type manager (as any other plugin manager) will throw a checked `PluginNotFoundException` if it is asked to act on an entity type that it doesn't know about. The idea is that is should be handled by the calling code: they can provide fallbacks (by implementing `FallbackPluginManagerInterface`), ignore it if it is an optional plugin, or show relevant information to the user (for example an error message, or a form that allows to create or buy the plugin).

Here is an illustration of this principle in action, from `ContentModerationConfigureEntityTypesForm`:

```
  public function buildForm(array $form, FormStateInterface $form_state, WorkflowInterface $workflow = NULL, $entity_type_id = NULL) {
    try {
      $this->entityType = $this->entityTypeManager->getDefinition($entity_type_id);
    }
    catch (PluginNotFoundException $e) {
      throw new NotFoundHttpException();
    }
    // ...
  }
```

This form gets an entity type ID as user input, passes it on to the entity type manager, which throws a checked exception if the plugin type doesn't exist. This checked exception is caught by the calling code who knows how to deal with it, and converted into a `NotFoundHttpException` which is (you guessed it) an unchecked exception which extends `\RuntimeException`.

Checked exceptions that are not handled are indicative of programming errors. They should either be checked (caught), or documented on all the public methods that possibly throw them, so that it is clear to the calling code that _they_ are responsible for catching them. In this particular example the calling code is the form builder which has no knowledge of the entity type manager, so it cannot handle it. That means this code _must_ catch it. It does and decides to let this bubble up as an HTTP error to the end user.

**Handling errors that "will never be thrown"**

Here is some information written by our colleague Andreas on the matter:
https://stackoverflow.com/questions/35586882/php-or-other-strategy-to-deal-with-exceptions-that-cannot-occur

So what is the point of converting this `PluginNotFoundException` into a `RuntimeException` if the `PluginNotFoundException` can never be thrown? I agree that on the surface it can seem pointless and not bringing value. But as explained the actual value is in the handling (catching) of the checked exception, not in the rethrowing.

So why do I do this anyway? Let's have a look at typical ways to handle "impossible" checked exceptions:

*The empty catch block*

```
try {
  $definition = $this->entityTypeManager->getDefinition('webtools_analytics_rule');
}
catch (PluginNotFoundException $e) {
  // This exception not occur.
}
```

The simplest way to deal with checked exceptions is to have an empty `catch {}` block. This is very commonly used. In many use cases this is fine, since the exception is not supposed to happen, right?

But there is a catch in highly complex systems like Drupal (pun intended). It is not unheard of that this "impossible" exception still happens in some rare edge case. For example the entity type manager might rely on data cached in Redis. At some point the Redis server gets out of sync due to a bug unrelated to our code, and suddenly the entity type manager thinks that our entity type doesn't exist and throws the "impossible" exception. If we ignore it completely then we have no trace of the runtime problem that occurred.

*Logging*

```
try {
  $definition = $this->entityTypeManager->getDefinition('webtools_analytics_rule');
}
catch (PluginNotFoundException $e) {
    $this->logger->warning('The webtools_analytics_rule entity type does not exist.');
}
```

One common way of still keeping track of "impossible" exceptions that are being thrown is to log them. This has some overhead of having to inject the logger service.

*Rethrowing as an unchecked exception*

```
try {
  $definition = $this->entityTypeManager->getDefinition('webtools_analytics_rule');
}
catch (PluginNotFoundException $e) {
  throw new \RuntimeException('The webtools_analytics_rule entity type does not exist.', 0, $e);
}
```

The third possibility is to simply rethrow the "impossible" exception as an unchecked `\RuntimeException`. This clearly communicates to the calling code that if this goes wrong it is because of a runtime problem, and we still get the advantages of Drupal's error handling that will log the error and show a message to the user.

So basically converting the checked exception into an unchecked exception is a convenient and robust way to handle this "exceptional" situation.